mlflow/sagemaker/__init__.py

```diff
@@ -316,36 +388,121 @@ def _upload_s3(local_model_path, bucket, prefix):
+    while (not endpoint_found) and ("NextToken" in deployed_endpoints_page):
+        next_token = deployed_endpoints_page["NextToken"]
+        deployed_endpoints_page = sage_client.list_endpoints(MaxResults=100, NextToken=next_token)
+        for endp in deployed_endpoints_page["Endpoints"]:
+            if endp["EndpointName"] == app_name:
+                endpoint_found = True
+                break
+
+    if endpoint_found and mode == DEPLOYMENT_MODE_CREATE:
+        msg = ("You are attempting to deploy an application with name: `{an}` in `{mcr} `mode." 
+               " However, an application with the same name already exists. If you want to update" 
+               " this application, deploy in `{madd}` or `{mrep}` mode.").format(
+                       an=app_name,
+                       mcr=DEPLOYMENT_MODE_CREATE,
+                       madd=DEPLOYMENT_MODE_ADD,
+                       mrep=DEPLOYMENT_MODE_REPLACE)
+        
+        raise Exception(msg)
+    
+    elif endpoint_found:
+        return _update_sagemaker_endpoint(endpoint_name=app_name,
+                                          image_url=image_url,
+                                          model_s3_path=model_s3_path,
+                                          run_id=run_id,
+                                          mode=mode, 
+                                          archive=archive,
+                                          role=role,
+                                          sage_client=sage_client,
+                                          s3_client=s3_client)
+
+    else:
+        return _create_sagemaker_endpoint(endpoint_name=app_name,
+                                          image_url=image_url,
+                                          model_s3_path=model_s3_path,
+                                          run_id=run_id,
+                                          role=role,
+                                          sage_client=sage_client)
+
+
+def _get_sagemaker_resource_unique_id():
+    """
+    :return: A unique, identifier that can be appended to a user-readable 
+    resource name to avoid naming collisions.
+    """
+    uuid_bytes = uuid.uuid4().bytes
+    # Use base64 encoding to shorten the UUID length. Note that the replacement of the 
+    # unsupported '+' symbol maintains uniqueness because the UUID byte string is of a fixed,
+    # 32-byte length
+    uuid_b64 = uuid_bytes.encode("base64").rstrip('=\n').replace("/", "-").replace("+", "AB")
+    return uuid_b64
+
+def _get_sagemaker_model_name(endpoint_name):
+    unique_id = _get_sagemaker_resource_unique_id()
+    return "{en}-model-{uid}".format(en=endpoint_name,
+                                     uid=unique_id)
+
+def _get_sagemaker_config_name(endpoint_name):
+    unique_id = _get_sagemaker_resource_unique_id()
+    return "{en}-config-{uid}".format(en=endpoint_name,
+                                      uid=unique_id)
+
+def _create_sagemaker_endpoint(endpoint_name, 
```

Comment:
nit: pep8 (though we may have given up on that in this repo)

Reply:
Fixed - turns out that this all fits on one line.